idseq_dag/util/count.py

```diff
@@ -88,7 +79,100 @@ def files_have_min_reads(input_files, min_reads):
     return True
+
+def _count_reads_expanding_duplicates(local_file_path, cluster_sizes):
+    # See documentation for reads_in_group use case with cluster_sizes, below.
+    unique_count, nonunique_count = 0, 0
+    for read in fasta.iterator(local_file_path):
+        # A read header looks someting like
+        #
+        #    >M05295:357:000000000-CRPNR:1:1101:22051:10534 OPTIONAL RANDOM STUFF"
+        #     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+        #
+        # where the first character on the line is '>' and the read ID (underlined above)
+        # extends from '>' to the first whitespace character, not including '>' itself.
+        #
+        # The fasta iterator already asserts that read.header[0] is '>'.
+        #
+        read_id, _ = read.header.split(None, 1)[1:]
+        unique_count += 1
+        nonunique_count += get_read_dcr(cluster_sizes, read_id)
+    return unique_count, nonunique_count
+
+def reads(local_file_path):
+    '''
+    Count reads in given FASTA or FASTQ file.
+    Implemented via wc, so very fast.
+    '''
+    return _count_reads_via_wc(local_file_path, max_reads=None)
+
+def reads_in_group(file_group, max_fragments=None, cluster_sizes=None):
+    '''
+    OVERVIEW
+
+    Count reads in a group of matching files, up to a maximum number of fragments,
+    optionally expanding each fragment to its cdhit cluster size.  Inputs in
+    FASTA and FASTQ format are supported, subject to restrictions below.
+
+    DEFINITIONS
+
+    The term "fragment" refers to the physical DNA fragments the reads derive from.
+    If the input is single, then 1 fragment == 1 read.
+    If the input is paired, then 1 fragment == 2 reads.
+
+    The term "cluster" refers to the output of the pipeline step cd-hit-dup, which
+    identifies and groups together duplicate fragments into clusters.  Duplicates
+    typically result from PCR or other amplificaiton.  It is cost effective for the
+    pipeline to operate on a single representative fragment from each cluster, then
+    infer back the original fragment count using the cluster sizes information
+    emitted by the cd-hit-dup step.  The caller should load that information via
+    m8.load_cluster_sizes() before passing it to this function.
+
+    RESTRICTIONS
+
+    When cluster_sizes is specified, the input must be in FASTA format, and the max_reads
+    argument must be unspecified (None).  Thus, the caller should choose between two
+    distinct use cases:
+
+       (*) Expand clusters when counting FASTA fragments.
+
+       (*) Count FASTA or FASTQ fragments without expanding clusters,
+           optionally truncating to max_fragments.  The input may
+           even be gz compressed.
+
+    PERFORMANCE
+
+    When cluster_sizes is not specified, the implementation is very fast, via wc.
+
+    When cluster_sizes is specified, the python implementation can process only about
+    3-4 million reads per minute. Fortunately, only steps run_lzw and run_bowtie2 can
```

Comment:
good to see you benchmarked. how much total runtime do you expect this to add to a typical run?

Reply:
for most runs, less than 15 seconds total